### PR TITLE
plugin/kubernetes: Add wildcards option

### PR DESCRIPTION
**Pull request by chrisohaver**
### 1. Why is this pull request needed and what does it do?

Add option to enable/disable wildcard queries. Default for now is `enabled`.

### 2. Which issues (if any) are related?

#4984

### 3. Which documentation changes (if any) need to be made?

included

### 4. Does this introduce a backward incompatible change or deprecation?

No.  But if we default to `disabled`, it would be.

### CODE REVIEW

**Comment by chrisohaver** [comment]
IMO, it would be nice if we could remove the feature entirely.  But if we do, we should give users a very long runway.
So, for now default to `enabled`, and make it an option to disable.
Next minor release perhaps, make it default to `disabled`.
In the future, maybe we can remove wildcard query support completely.

**Comment by miekg** [comment]
Read the issue. Agree with Chris this should be removed. As it's
intertwined with etcd we should remove it there as well IMO.

Probably a knob know default to true, then false and the remove the code?

File: plugin/kubernetes/README.md [hunk 1]
```diff
@@ -101,7 +102,9 @@ kubernetes [ZONES...] {
 * `ignore empty_service` returns NXDOMAIN for services without any ready endpoint addresses (e.g., ready pods).
   This allows the querying pod to continue searching for the service in the search path.
   The search path could, for example, include another Kubernetes cluster.
-
+* `wildcards enabled|disabled` enables or disables the wildcard query feature (see Wildcards section below).
+  Default is enabled.
```
Comment (miekg) [inline plugin/kubernetes/README.md:106]: Hmm, another option that requires deprecating. I wonder can't we just rip the wildcard stuff out at once and be done with it?

**Comment by chrisohaver** [comment]
> Hmm, another option that requires deprecating. I wonder can't we just rip the wildcard stuff out at once and be done with it?

This is the user friendly / long runway approach, which I tend to gravitate to.  We could also just remove it - the maintainer friendly approach.

**Comment by miekg** [comment]
[ Quoting ***@***.***> in "Re: [coredns/coredns] plugin/kubern..." ]
>This is the user friendly / long runway approach, which I tend to gravitate to.
>We could also just remove it - the maintainer friendly approach.

I've seen people use it on twitter, as nice debug feature, but we have axfr now if you
want this. I'm inclined to say deprecate this in one go, and be done with it. DNS as such
can use it; you specifically need to craft a query to use this.

**Comment by chrisohaver** [comment]
>... As it's intertwined with etcd we should remove it there as well IMO. 

Wildcard behavior is part of SkyDNS, which etcd emulates ...

https://github.com/skynetservices/skydns#wildcards

... So I'm not sure if we should remove it from the etcd plugin.

**Comment by johnbelamaric** [comment]
yeah, seems if someone is using that plugin it's probably what they want. if it's not, perhaps a *new* etcd plugin is needed that doesn't use that functionality

**Comment by chrisohaver** [comment]
We could skip adding an option, deprecate wildcards now and  remove them in next release with #5019 + #5030.

That said, the advantage of adding an option (this PR) is that users could disable wildcards before waiting for the next minor release, if they are concerned about the information disclosure (#4984).

**Comment by chrisohaver** [comment]
Closing this.  Just deprecating and removing wildcards is simpler.  In the mean time, users can disable wildcard queries in kubernetes using the suggestions outlined in https://github.com/coredns/coredns/issues/4984#issuecomment-969272093.

**Pull request closed without merging**